cloud/edgecontroller/pkg/devicecontroller/apis/devices/v1alpha1/device_model_types.go

```diff
@@ -0,0 +1,168 @@
+	// Required: The device property name to be accessed. This should refer to one of the
+	// device properties defined in the device model.
+	PropertyName string `json:"propertyName,omitempty"`
+	// Required: Protocol relevant config details about the how to access the device property.
+	VisitorConfig `json:",inline"`
+}
+
+// At least one of its members must be specified.
+type VisitorConfig struct {
+	// Opcua represents a set of additional visitor config fields of opc-ua protocol.
+	// +optional
+	OpcUA VisitorConfigOPCUA `json:"opcua,omitempty"`
+	// Modbus represents a set of additional visitor config fields of modbus protocol.
+	// +optional
+	Modbus VisitorConfigModbus `json:"modbus,omitempty"`
+}
+
+// Common visitor configurations for opc-ua protocol
+type VisitorConfigOPCUA struct {
+	// Required: The ID of opc-ua node, e.g. "ns=1,i=1005"
+	NodeID string `json:"nodeID,omitempty"`
+	// The name of opc-ua node
+	BrowseName string `json:"browseName,omitempty"`
+}
+
+// Common visitor configurations for modbus protocol
+type VisitorConfigModbus struct {
+	// Required: Type of register
+	Register ModbusRegisterType `json:"register,omitempty"`
+	// Required: Offset indicates the starting register number to read/write data.
+	Offset int64 `json:"offset,omitempty"`
+	// Required: Limit number of registers to read/write.
+	Limit int64 `json:"limit,omitempty"`
+	// The scale to convert raw property data into final units.
+	// Defaults to 1.0
+	// +optional
+	Scale float64 `json:"scale,omitempty"`
+	// Indicates whether the high and low byte swapped.
+	// Defaults to false.
+	// +optional
+	IsSwap bool `json:"isSwap,omitempty"`
+	// Indicates whether the high and low register swapped.
+	// Defaults to false.
+	// +optional
+	IsRegisterSwap bool `json:"isRegisterSwap,omitempty"`
+}
+
+// The Modbus register type to read a device property.
+type ModbusRegisterType string
+
+// Modbus protocol register types
+const (
+	ModbusRegisterTypeCoilRegister          ModbusRegisterType = "CoilRegister"
+	ModbusRegisterTypeDiscreteInputRegister ModbusRegisterType = "DiscreteInputRegister"
+	ModbusRegisterTypeInputRegister         ModbusRegisterType = "InputRegister"
+	ModbusRegisterTypeHoldingRegister       ModbusRegisterType = "HoldingRegister"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// DeviceModel is the Schema for the device model API
+// +k8s:openapi-gen=true
+type DeviceModel struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec DeviceModelSpec `json:"spec,omitempty"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// DeviceModelList contains a list of DeviceModel
+type DeviceModelList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []DeviceModel `json:"items"`
+}
+
+func init() {
+	//SchemeBuilder.Register(&Device{}, &DeviceList{}, &DeviceModel{})
```

Comment:
We are not registering device and devicemodel here?

Reply:
no, these generated files require k8s version 1.11 or above, so we will register them in the controller code.

Reply:
this code depends on   "sigs.k8s.io/controller-runtime" project which has k8s dependency >= 1.11.
our current k8s dependency is 1.10.9, so we can't use the controller-runtime package for scheme related operations. we are doing this manually as Sid mentioned

Reply:
So we can remove the `init()` here?